### PR TITLE
MAINT: use lapack for scipy.linalg.norm

### PR DESCRIPTION
**Pull request by argriffing**
Use the `*lange` functions wrapped in https://github.com/scipy/scipy/pull/4603 for `scipy.linalg.norm`.  The scipy matrix norms appear to not have been tested (only vector norms) so I've added some tests.  I've also attempted to add `asv` benchmarks for the matrix `1-norm` and the matrix `inf-norm`.

### CODE REVIEW

**Comment by ewmoore** [comment]
Seems kind of awful that this will copy a C order array to F order just to find the norm.

**Comment by argriffing** [comment]
@ewmoore Yes, maybe this code would be improved by checking that the matrix is more strictly compatible with lapack, otherwise using the default path (calling the numpy linalg norm)? Also if the distinction between the orderings just corresponds to a transposition, and is the matrix is contiguous, then this could possibly be detected and handled.  For example I think the 1-norm of M is the inf-norm of the transpose of M, and the Frobenius norm is invariant to transposition of M.  I've been having trouble using `asv` through `runtests.py` so I don't yet have the benchmark results that I'd like.  Another possible slowness is that I think the input is always checked for finiteness, although I would guess that this would not cause as much slowness as making extra copies.

**Comment by insertinterestingnamehere** [comment]
+1 to using the transposes to get 1 and infinity norms of C contiguous arrays.

**Comment by argriffing** [comment]
The benchmark results show a performance improvement for small matrices and a performance regression for larger matrices.  Only square matrices have been included in the benchmarks.

Are there documented tricks for determining under what conditions a matrix will be copied, before calling a wrapped lapack function?  For example would the following make sense?
```python
if first_condition(M):
    return lapack_implementation_of_some_norm(M)
elif some_other_condition(M):
    return lapack_implementation_of_some_complementary_norm(M.T)
else:
    return default_numpy_implementation_of_the_norm(M)
```
What would be the appropriate conditions?  The lapack matrix norm function has something called `LDA` which seems to be a stride length.  Does this mean it can deal with non-contiguous matrices?  Does the numpy<->fortran interface deal with this automatically?

**Comment by argriffing** [comment]
After trying to be clever about c/fortran ordering and array contiguity, this gives the following `asv` benchmarks comparison using
`$ python runtests.py --bench-compare master linalg.Bench`

```
BENCHMARK                                    BEFORE      AFTER     FACTOR
....time_inf_norm(20, 'nocont', 'scipy')    34.30μs    40.13μs   1.17009161x
....time_inf_norm(20, 'nocont', 'numpy')    19.66μs    21.96μs   1.11683783x
...ch.time_1_norm(20, 'nocont', 'scipy')    33.81μs    36.97μs   1.09333703x
...ench.time_det(500, 'contig', 'numpy')     9.38ms    10.14ms   1.08162029x
...time_inf_norm(100, 'nocont', 'scipy')    75.17μs    80.97μs   1.07723861x
...ench.time_svd(100, 'nocont', 'scipy')     5.12ms     5.49ms   1.07343795x
...ench.time_svd(100, 'contig', 'scipy')     5.13ms     5.50ms   1.07118606x
...ench.time_svd(100, 'contig', 'numpy')     5.16ms     5.53ms   1.07113718x
...ch.time_solve(100, 'nocont', 'numpy')   222.55μs   238.00μs   1.06944344x
...ench.time_det(500, 'contig', 'scipy')     9.58ms    10.21ms   1.06569122x
...ch.time_1_norm(20, 'contig', 'numpy')    17.58μs    18.67μs   1.06201930x
...time_eigvals(1000, 'nocont', 'scipy')      1.43s      1.52s   1.05975533x
...h.time_1_norm(100, 'nocont', 'scipy')    71.70μs    75.48μs   1.05271927x
....time_1_norm(1000, 'nocont', 'numpy')     2.09ms     1.99ms   0.94864103x
...Bench.time_inv(20, 'contig', 'numpy')    35.94μs    34.03μs   0.94663780x
...Bench.time_det(20, 'contig', 'numpy')    27.65μs    26.11μs   0.94440608x
....time_eigvals(500, 'nocont', 'numpy')   291.06ms   271.91ms   0.93418938x
....time_eigvals(500, 'nocont', 'scipy')   289.60ms   270.41ms   0.93375183x
...time_inf_norm(500, 'nocont', 'scipy')   848.98μs   782.27μs   0.92142588x
...ench.time_det(100, 'contig', 'scipy')   244.99μs   223.70μs   0.91310892x
...time_inf_norm(500, 'nocont', 'numpy')   560.31μs   509.44μs   0.90921027x
...Bench.time_inv(20, 'nocont', 'numpy')    37.44μs    33.97μs   0.90727052x
...Bench.time_inv(20, 'contig', 'scipy')    79.71μs    70.27μs   0.88166773x
....time_eigvals(100, 'contig', 'scipy')     8.11ms     7.14ms   0.87950742x
...h.time_1_norm(500, 'contig', 'scipy')   646.81μs   541.62μs   0.83737324x
...time_inf_norm(500, 'contig', 'scipy')   748.16μs   614.02μs   0.82070655x
...ime_inf_norm(1000, 'contig', 'numpy')     2.80ms     2.27ms   0.81010681x
...time_inf_norm(100, 'contig', 'scipy')    66.92μs    46.05μs   0.68822453x
....time_1_norm(1000, 'contig', 'scipy')     3.05ms     2.04ms   0.66876933x
...h.time_1_norm(100, 'contig', 'scipy')    65.03μs    42.26μs   0.64982718x
...ime_inf_norm(1000, 'contig', 'scipy')     3.65ms     2.36ms   0.64485130x
....time_inf_norm(20, 'contig', 'scipy')    30.56μs    16.54μs   0.54110413x
...ch.time_1_norm(20, 'contig', 'scipy')    29.30μs    15.81μs   0.53948688x
```

**Comment by argriffing** [comment]
I've updated the matrix norm benchmarks, now including rectangular matrices with negative entries allowed.  Benchmarks for the Frobenius matrix norm have been added, and the norm benchmarks are now grouped into their own benchmark class.  For some reason `*lange` introduces a ~3x performance regression for the Frobenius norm, so I'm keeping the old implementation (deferring to numpy for that matrix norm).
`$ python runtests.py --bench-compare master linalg.Norm`

```
BENCHMARK                                    BEFORE      AFTER     FACTOR
...inf_norm((20, 20), 'nocont', 'scipy')    34.25μs    38.84μs   1.13421128x
...norm((1000, 1000), 'contig', 'scipy')     3.42ms     3.81ms   1.11510672x
...ius_norm((20, 20), 'nocont', 'scipy')    25.52μs    27.90μs   1.09328575x
...f_norm((1000, 20), 'nocont', 'scipy')   195.27μs   212.73μs   1.08939390x
...e_1_norm((20, 20), 'nocont', 'scipy')    34.38μs    36.93μs   1.07415000x
...ius_norm((20, 20), 'contig', 'scipy')    22.72μs    24.14μs   1.06245835x
...s_norm((1000, 20), 'nocont', 'scipy')   103.35μs    97.59μs   0.94427535x
...e_1_norm((20, 20), 'contig', 'numpy')    18.18μs    17.12μs   0.94169906x
...s_norm((1000, 20), 'nocont', 'numpy')    64.00μs    59.59μs   0.93101985x
...norm((1000, 1000), 'nocont', 'numpy')     6.49ms     6.00ms   0.92440025x
...1_norm((20, 1000), 'contig', 'scipy')    94.46μs    71.45μs   0.75633468x
...f_norm((20, 1000), 'contig', 'scipy')    99.01μs    73.06μs   0.73787171x
...norm((1000, 1000), 'contig', 'scipy')     3.44ms     2.34ms   0.67933971x
...norm((1000, 1000), 'contig', 'scipy')     3.02ms     2.05ms   0.67865180x
...f_norm((100, 100), 'contig', 'scipy')    67.74μs    44.78μs   0.66106087x
...1_norm((100, 100), 'contig', 'scipy')    66.04μs    42.79μs   0.64792731x
...1_norm((1000, 20), 'contig', 'scipy')   104.38μs    67.61μs   0.64771496x
...f_norm((1000, 20), 'contig', 'scipy')   112.48μs    71.63μs   0.63682450x
...inf_norm((20, 20), 'contig', 'scipy')    29.43μs    17.09μs   0.58074653x
...e_1_norm((20, 20), 'contig', 'scipy')    29.01μs    16.38μs   0.56482780x
```

**Comment by insertinterestingnamehere** [comment]
It is doing scaling similar to what is done with `*nrm2` to avoid floating point overflow. See http://www.netlib.org/lapack/explore-html/dc/d09/dlange_8f_source.html (line 200). The Frobenius norm is done using a call to `*lassq`, which does scaled arithmetic (like `*nrm2` does).

For consistency's sake, eventually we should either use scaled arithmetic in every case where it's possible (even along axes of multidimensional arrays) or require users to use the LAPACK wrappers directly when they need the scaled versions. As it stands, the less-reliable version is used in all cases except when taking the 2-norm of a 1D array. That's not very consistent. Hiding the scaled arithmetic the way we currently do lends itself to strange speed results.
For example (informal timings from ipython with current master):

    import numpy as np
    from scipy.linalg import norm
    a = np.random.rand(10000)
    norm(a)  # 22.4 µs
    norm(a[None], 'fro')  # 42.7 µs
    np.linalg.norm(a)  # 7.44 µs

and this lovely example:

    a = np.random.rand(1000, 1000)
    norm(a, 'fro')  # 4.59 ms
    np.linalg.norm(a, 'fro') # 3.43 ms
    norm(a.ravel())  # 1.46 ms
    np.linalg.norm(a.ravel())  # 396 µs

Oddly enough, the difference between the Frobenius norm and the 2-norm is present in the unscaled numpy version as well. I'm not sure how to explain that one.

**Comment by insertinterestingnamehere** [comment]
Really, this PR is a net improvement. The 2-norm/frobenius timing inconsistencies should be taken care of separately anyway. LGTM.

**Comment by argriffing** [comment]
> and this lovely example

In `numpy` this strange distinction seems to be a consequence of the complication of the `norm` code which handles vector norms, matrix norms, various orders and formulas of norms, functions that aren't technically norms, and `axis` and `keepdims` arguments. The `np.linalg.norm(a.ravel())` call takes a short code path because it has no optional arguments, whereas `np.linalg.norm(a, 'fro')` takes the fully general code path because it has the extra argument; `np.linalg.norm(a)` is fast for me and also computes the Frobenius norm.

**Comment by argriffing** [comment]
Added a numpy PR https://github.com/numpy/numpy/pull/5701 to address its inconsistencies between speeds of `norm(A)` and `norm(A, 'fro')` which compute the same thing when A is a 2d array.  After reading @insertinterestingnamehere's comments I now understand the point about the scaling.  I guess the idea is that `*lange` is more clever about computing things like `sqrt(sum(square(x)))` without losing precision, kind of analogously to `logsumexp(x)` vs. `log(sum(exp(x)))`.

**Comment by argriffing** [comment]
> Seems kind of awful that this will copy a C order array to F order just to find the norm.

OK I think this PR avoids doing that now.

**Comment by insertinterestingnamehere** [comment]
@argriffing Thanks for taking care of that. This looks good to me.

File: scipy/linalg/misc.py [hunk 1]
```diff
@@ -112,14 +113,34 @@ def norm(a, ord=None):
...
-    if ord in (None, 2) and (a.ndim == 1) and (a.dtype.char in 'fdFD'):
-        # use blas for fast and stable euclidean norm
-        func_name = _nrm2_prefix.get(a.dtype.char, 'd') + 'nrm2'
-        nrm2 = getattr(blas, func_name)
-        return nrm2(a)
+    if a.dtype.char in 'fdFD':
+        if ord in (None, 2) and (a.ndim == 1):
+            # use blas for fast and stable euclidean norm
+            func_name = _nrm2_prefix.get(a.dtype.char, 'd') + 'nrm2'
+            nrm2 = getattr(blas, func_name)
```
Comment (ev-br) [inline scipy/linalg/misc.py]: looks like `get_blas_funcs`

**Comment by argriffing** [comment]
@ev-br Thanks, I've made those changes; waiting for TravisCI...

**Comment by ev-br** [comment]
Thanks @argriffing, green button it is.

**Pull request merged**